source/client/benchmark_client_impl.cc

```diff
@@ -21,11 +21,18 @@ using namespace std::chrono_literals;
 namespace Nighthawk {
 namespace Client {
 
-void Http1PoolImpl::createConnections(const uint32_t connection_limit) {
-  ENVOY_LOG(error, "Prefetching {} connections.", connection_limit);
-  for (uint32_t i = 0; i < connection_limit; i++) {
-    createNewConnection();
+Envoy::Http::ConnectionPool::Cancellable*
+Http1PoolImpl::newStream(Envoy::Http::StreamDecoder& response_decoder,
+                         Envoy::Http::ConnectionPool::Callbacks& callbacks) {
+  // In prefetch mode we try to keep the amount of connections at the configured limit.
+  if (prefetch_connections_) {
+    while (host_->cluster().resourceManager(priority_).connections().canCreate()) {
```

Comment:
What guarantee do we have that this doesn't live lock?

Reply:
So.. how would you see this lock? 
Note: each worker in NH has a unique cluster assigned to it / they are not shared across workers. As far as I can tell, the vanilla Envoy pool implementations also use this in vanilla Envoy.

Reply:
I'm wondering what guarntees we have that `canCreate()` returns false at some point?

Reply:
The call to `ConnPoolImpl::createNewConnection()` unconditionally constructs a new `ActiveClient`, which in turn unconditionally and directly calls `resourceManager(parent_.priority_).connections().inc();`.

The resource manager is configured to cap the number of connections to our `--connection` limit option, and at some point we're going to hit that at which point `canCreate()` will return false.